src/ethereum_watcher.rs

```diff
@@ -1,43 +1,531 @@
+                &action_sender,
+                watch_config.block_production_alert.alert_action.clone(),
+                Some(watch_config.block_production_alert.alert_level.clone()),
+            );
+            return;
+        }
+    };
+
+    if seconds_since_last_block > watch_config.block_production_alert.max_block_time {
+        send_alert(
+            &alert_sender,
+                String::from("Ethereum block is taking long"),
+        format!(
+                "Next ethereum block is taking longer than {} seconds. Last block was {} seconds ago.",
+                watch_config.block_production_alert.max_block_time, seconds_since_last_block
+            ),
+            watch_config.block_production_alert.alert_level.clone(),
+        );
+        send_action(
+            &action_sender,
+            watch_config.block_production_alert.alert_action.clone(),
+            Some(watch_config.block_production_alert.alert_level.clone()),
+        );
+    }
+}
+
+async fn check_account_balance(
+    ethereum_chain: Arc<dyn EthereumChainTrait>,
+    action_sender: UnboundedSender<ActionParams>,
+    alert_sender: UnboundedSender<AlertParams>,
+    watch_config: &EthereumClientWatcher,
+    account_address: &Option<String>,
+) {
+
+    // Return early if there's no address or if the alert level is None.
+    let address = match account_address {
+        Some(addr) => addr,
+        None => return,
+    };
+
+    if watch_config.account_funds_alert.alert_level == AlertLevel::None {
+        return;
+    }
+
+    // Proceed with checking the account balance
+    let retrieved_balance = match ethereum_chain.get_account_balance(address).await {
+        Ok(balance) => balance,
+        Err(e) => {
+            send_alert(
+                &alert_sender,
+                String::from("Failed to check ethereum account funds"),
+                format!("Failed to check ethereum account funds: {}", e),
+                watch_config.account_funds_alert.alert_level.clone(),
+            );
+            send_action(
+                &action_sender,
+                watch_config.account_funds_alert.alert_action.clone(),
+                Some(watch_config.account_funds_alert.alert_level.clone()),
+            );
+            return;
+        }
+    };
+
+    let min_balance = get_value(
+        watch_config.account_funds_alert.min_balance,
+        18,
+    );
+    if retrieved_balance < min_balance {
+        send_alert(
+            &alert_sender,
+            String::from("Ethereum account low on funds"),
+            format!(
+                "Ethereum account ({}) is low on funds. Current balance: {}",
+                address, retrieved_balance,
+            ),
+            watch_config.account_funds_alert.alert_level.clone(),
+        );
+        send_action(
+            &action_sender,
+            watch_config.account_funds_alert.alert_action.clone(),
+            Some(watch_config.account_funds_alert.alert_level.clone()),
+        );
+    }
```

Comment:
is this like a relayer that needs to have eth to send txes?

Reply:
Yes, send_action -> sends an action to the ethereum_actions.rs which will then execute an action: Pause on one or all the contracts, depends on configuration.